File: android/engine/src/main/kotlin/io/sauti/engine/Backoff.kt
package io.sauti.engine

import kotlin.math.min
import kotlin.math.roundToLong

fun computeBackoff(
    attempt: Int,
    baseMs: Long,
    maxMs: Long,
    random: () -> Double
): Long {
    val exponential = baseMs.toDouble() * Math.pow(2.0, attempt.toDouble())
    val capped = min(exponential, maxMs.toDouble())
    val jitter = capped * 0.25 * random()
    return (capped - capped * 0.25 + jitter).roundToLong()
}

File: packages/core/src/backoff.ts
export function computeBackoff(
  attempt: number,
  baseMs: number,
  maxMs: number,
  random: () => number = Math.random
): number {
  const exponential = baseMs * 2 ** attempt;
  const capped = Math.min(exponential, maxMs);
  const jitter = capped * 0.25 * random();
  return Math.round(capped - capped * 0.25 + jitter);
}
